File: ts/src/ast.ts
import type { SourceSpan } from './token.js';

// ── Expressions ──

export type ElwoodExpression =
  | PipelineExpression
  | PathExpression
  | IdentifierExpression
  | LambdaExpression
  | ObjectExpression
  | ArrayExpression
  | LiteralExpression
  | InterpolatedStringExpression
  | BinaryExpression
  | UnaryExpression
  | IfExpression
  | MatchExpression
  | MemoExpression
  | MethodCallExpression
  | FunctionCallExpression
  | MemberAccessExpression
  | IndexExpression;

// ── Top-level ──

export interface ScriptNode {
  type: 'Script';
  bindings: LetBindingNode[];
  returnExpression: ElwoodExpression | null;
  span: SourceSpan;
}

export interface LetBindingNode {
  type: 'LetBinding';
  name: string;
  value: ElwoodExpression;
  span: SourceSpan;
}

// ── Expression nodes ──

export interface PipelineExpression {
  type: 'Pipeline';
  source: ElwoodExpression;
  operations: PipeOperation[];
  span: SourceSpan;
}

export interface PathExpression {
  type: 'Path';
  segments: PathSegment[];
  isRooted: boolean;
  span: SourceSpan;
}

export interface IdentifierExpression {
  type: 'Identifier';
  name: string;
  span: SourceSpan;
}

export interface LambdaExpression {
  type: 'Lambda';
  parameters: string[];
  body: ElwoodExpression;
  span: SourceSpan;
}

export interface ObjectExpression {
  type: 'Object';
  properties: ObjectProperty[];
  span: SourceSpan;
}

export interface ObjectProperty {
  key: string;
  value: ElwoodExpression;
  span: SourceSpan;
  isSpread?: boolean;
  computedKey?: ElwoodExpression;
}

export interface ArrayExpression {
  type: 'Array';
  items: ElwoodExpression[];
  span: SourceSpan;
}

export interface LiteralExpression {
  type: 'Literal';
  value: string | number | boolean | null;
  span: SourceSpan;
}

export interface InterpolatedStringExpression {
  type: 'InterpolatedString';
  parts: InterpolationPart[];
  span: SourceSpan;
}

export type InterpolationPart =
  | { type: 'Text'; text: string; span: SourceSpan }
  | { type: 'Expression'; expression: ElwoodExpression; span: SourceSpan };

export interface BinaryExpression {
  type: 'Binary';
  left: ElwoodExpression;
  operator: BinaryOperator;
  right: ElwoodExpression;
  span: SourceSpan;
}

export interface UnaryExpression {
  type: 'Unary';
  operator: UnaryOperator;
  operand: ElwoodExpression;
  span: SourceSpan;
}

export interface IfExpression {
  type: 'If';
  condition: ElwoodExpression;
  thenBranch: ElwoodExpression;
  elseBranch: ElwoodExpression;
  span: SourceSpan;
}

export interface MatchExpression {
  type: 'Match';
  input: ElwoodExpression;
  arms: MatchArm[];
  span: SourceSpan;
}

export interface MatchArm {
  pattern: ElwoodExpression | null; // null = wildcard
  result: ElwoodExpression;
  span: SourceSpan;
}

export interface MemoExpression {
  type: 'Memo';
  lambda: LambdaExpression;
  span: SourceSpan;
}

export interface MethodCallExpression {
  type: 'MethodCall';
  target: ElwoodExpression;
  methodName: string;
  arguments: ElwoodExpression[];
  span: SourceSpan;
}

export interface FunctionCallExpression {
  type: 'FunctionCall';
  functionName: string;
  arguments: ElwoodExpression[];
  span: SourceSpan;
}

export interface MemberAccessExpression {
  type: 'MemberAccess';
  target: ElwoodExpression;
  memberName: string;
  span: SourceSpan;
}

export interface IndexExpression {
  type: 'Index';
  target: ElwoodExpression;
  index: ElwoodExpression | null; // null = [*]
  span: SourceSpan;
}

// ── Pipe Operations ──

export type PipeOperation =
  | WhereOperation
  | SelectOperation
  | SelectManyOperation
  | OrderByOperation
  | GroupByOperation
  | DistinctOperation
  | AggregateOperation
  | SliceOperation
  | TakeWhileOperation
  | BatchOperation
  | JoinOperation
  | ConcatOperation
  | ReduceOperation
  | QuantifierOperation
  | MatchOperation;

export interface WhereOperation { type: 'Where'; predicate: ElwoodExpression; span: SourceSpan }
export interface SelectOperation { type: 'Select'; projection: ElwoodExpression; span: SourceSpan }
export interface SelectManyOperation { type: 'SelectMany'; projection: ElwoodExpression; span: SourceSpan }
export interface OrderByOperation { type: 'OrderBy'; keys: { key: ElwoodExpression; ascending: boolean }[]; span: SourceSpan }
export interface GroupByOperation { type: 'GroupBy'; keySelector: ElwoodExpression; span: SourceSpan }
export interface DistinctOperation { type: 'Distinct'; span: SourceSpan }
export interface AggregateOperation { type: 'Aggregate'; name: string; predicate?: ElwoodExpression; span: SourceSpan }
export interface SliceOperation { type: 'Slice'; kind: 'take' | 'skip'; count: ElwoodExpression; span: SourceSpan }
export interface TakeWhileOperation { type: 'TakeWhile'; predicate: ElwoodExpression; span: SourceSpan }
export interface BatchOperation { type: 'Batch'; size: ElwoodExpression; span: SourceSpan }
export interface JoinOperation {
  type: 'Join';
  source: ElwoodExpression;
  leftKey: ElwoodExpression;
  rightKey: ElwoodExpression;
  intoAlias?: string;
  mode: JoinMode;
  span: SourceSpan;
}
export interface ConcatOperation { type: 'Concat'; separator?: ElwoodExpression; span: SourceSpan }
export interface ReduceOperation { type: 'Reduce'; accumulator: ElwoodExpression; initialValue?: ElwoodExpression; span: SourceSpan }
export interface QuantifierOperation { type: 'Quantifier'; kind: 'any' | 'all'; predicate: ElwoodExpression; span: SourceSpan }
export interface MatchOperation { type: 'MatchOp'; arms: MatchArm[]; span: SourceSpan }

export type JoinMode = 'inner' | 'left' | 'right' | 'full';

// ── Path Segments ──

export type PathSegment =
  | { type: 'Property'; name: string; span: SourceSpan }
  | { type: 'Index'; index: number | null; span: SourceSpan }  // null = wildcard [*]
  | { type: 'Slice'; start: number | null; end: number | null; span: SourceSpan }
  | { type: 'RecursiveDescent'; name: string; span: SourceSpan };

// ── Operators ──

export type BinaryOperator =
  | 'Add' | 'Subtract' | 'Multiply' | 'Divide'
  | 'Equal' | 'NotEqual'
  | 'LessThan' | 'LessThanOrEqual' | 'GreaterThan' | 'GreaterThanOrEqual'
  | 'And' | 'Or';

export type UnaryOperator = 'Negate' | 'Not';

File: dotnet/src/Elwood.Core/Syntax/Ast.cs
using Elwood.Core.Diagnostics;

namespace Elwood.Core.Syntax;

// ── Base ──

public abstract record ElwoodNode(SourceSpan Span);

// ── Top-level ──

/// <summary>A complete Elwood script: let bindings + optional return expression.</summary>
public sealed record ScriptNode(
    IReadOnlyList<LetBindingNode> Bindings,
    ElwoodExpression? ReturnExpression,
    SourceSpan Span
) : ElwoodNode(Span);

/// <summary>let name = expression</summary>
public sealed record LetBindingNode(
    string Name,
    ElwoodExpression Value,
    SourceSpan Span
) : ElwoodNode(Span);

// ── Expressions ──

public abstract record ElwoodExpression(SourceSpan Span) : ElwoodNode(Span);

/// <summary>A pipeline: expr | op1 | op2 | ...</summary>
public sealed record PipelineExpression(
    ElwoodExpression Source,
    IReadOnlyList<PipeOperation> Operations,
    SourceSpan Span
) : ElwoodExpression(Span);

/// <summary>JSONPath navigation: $.foo.bar[*].baz</summary>
public sealed record PathExpression(
    IReadOnlyList<PathSegment> Segments,
    bool IsRooted, // starts with $
    SourceSpan Span
) : ElwoodExpression(Span);

/// <summary>A named variable reference (from let binding or lambda parameter).</summary>
public sealed record IdentifierExpression(
    string Name,
    SourceSpan Span
) : ElwoodExpression(Span);

/// <summary>Lambda: x => expr  or  (x, y) => expr</summary>
public sealed record LambdaExpression(
    IReadOnlyList<string> Parameters,
    ElwoodExpression Body,
    SourceSpan Span
) : ElwoodExpression(Span);

/// <summary>Object literal: { key: expr, key2: expr2, ... }</summary>
public sealed record ObjectExpression(
    IReadOnlyList<ObjectProperty> Properties,
    SourceSpan Span
) : ElwoodExpression(Span);

public sealed record ObjectProperty(string Key, ElwoodExpression Value, SourceSpan Span, bool IsSpread = false, ElwoodExpression? ComputedKey = null);

/// <summary>Array literal: [expr, expr, ...]</summary>
public sealed record ArrayExpression(
    IReadOnlyList<ElwoodExpression> Items,
    SourceSpan Span
) : ElwoodExpression(Span);

/// <summary>String, number, bool, null literals.</summary>
public sealed record LiteralExpression(
    object? Value, // string, double, bool, or null
    SourceSpan Span
) : ElwoodExpression(Span);

/// <summary>String interpolation: `Hello {expr} world`</summary>
public sealed record InterpolatedStringExpression(
    IReadOnlyList<InterpolationPart> Parts,
    SourceSpan Span
) : ElwoodExpression(Span);

public abstract record InterpolationPart(SourceSpan Span);
public sealed record TextPart(string Text, SourceSpan Span) : InterpolationPart(Span);
public sealed record ExpressionPart(ElwoodExpression Expression, SourceSpan Span) : InterpolationPart(Span);

/// <summary>Binary expression: a + b, a > b, a && b</summary>
public sealed record BinaryExpression(
    ElwoodExpression Left,
    BinaryOperator Operator,
    ElwoodExpression Right,
    SourceSpan Span
) : ElwoodExpression(Span);

/// <summary>Unary expression: !a, -a</summary>
public sealed record UnaryExpression(
    UnaryOperator Operator,
    ElwoodExpression Operand,
    SourceSpan Span
) : ElwoodExpression(Span);

/// <summary>if cond then expr else expr</summary>
public sealed record IfExpression(
    ElwoodExpression Condition,
    ElwoodExpression ThenBranch,
    ElwoodExpression ElseBranch,
    SourceSpan Span
) : ElwoodExpression(Span);

/// <summary>expr | match { pattern => result, ... }</summary>
public sealed record MatchExpression(
    ElwoodExpression Input,
    IReadOnlyList<MatchArm> Arms,
    SourceSpan Span
) : ElwoodExpression(Span);

public sealed record MatchArm(
    ElwoodExpression? Pattern, // null = wildcard (_)
    ElwoodExpression Result,
    SourceSpan Span
);

/// <summary>Memoized lambda: memo params => body. Cached by argument values.</summary>
public sealed record MemoExpression(
    LambdaExpression Lambda,
    SourceSpan Span
) : ElwoodExpression(Span);

/// <summary>Method call on a value: expr.method(args)</summary>
public sealed record MethodCallExpression(
    ElwoodExpression Target,
    string MethodName,
    IReadOnlyList<ElwoodExpression> Arguments,
    SourceSpan Span
) : ElwoodExpression(Span);

/// <summary>Standalone function call: functionName(args)</summary>
public sealed record FunctionCallExpression(
    string FunctionName,
    IReadOnlyList<ElwoodExpression> Arguments,
    SourceSpan Span
) : ElwoodExpression(Span);

/// <summary>Property access on an expression: expr.property or expr?.property</summary>
public sealed record MemberAccessExpression(
    ElwoodExpression Target,
    string MemberName,
    SourceSpan Span,
    bool Optional = false
) : ElwoodExpression(Span);

/// <summary>Index access: expr[index] or expr[*]</summary>
public sealed record IndexExpression(
    ElwoodExpression Target,
    ElwoodExpression? Index, // null means [*] (all elements)
    SourceSpan Span
) : ElwoodExpression(Span);

// ── Pipe Operations ──

public abstract record PipeOperation(SourceSpan Span);

/// <summary>| where predicate</summary>
public sealed record WhereOperation(ElwoodExpression Predicate, SourceSpan Span) : PipeOperation(Span);

/// <summary>| select projection</summary>
public sealed record SelectOperation(ElwoodExpression Projection, SourceSpan Span) : PipeOperation(Span);

/// <summary>| selectMany projection</summary>
public sealed record SelectManyOperation(ElwoodExpression Projection, SourceSpan Span) : PipeOperation(Span);

/// <summary>| orderBy key [asc|desc]</summary>
public sealed record OrderByOperation(
    IReadOnlyList<(ElwoodExpression Key, bool Ascending)> Keys,
    SourceSpan Span
) : PipeOperation(Span);

/// <summary>| groupBy key</summary>
public sealed record GroupByOperation(ElwoodExpression KeySelector, SourceSpan Span) : PipeOperation(Span);

/// <summary>| distinct</summary>
public sealed record DistinctOperation(SourceSpan Span) : PipeOperation(Span);

/// <summary>| first, | last, | count, | sum, | min, | max</summary>
public sealed record AggregateOperation(string Name, SourceSpan Span, ElwoodExpression? Predicate = null) : PipeOperation(Span)
{
    public AggregateOperation(string name, ElwoodExpression predicate, SourceSpan span)
        : this(name, span, predicate) { }
}

/// <summary>| take n, | skip n</summary>
public sealed record SliceOperation(string Kind, ElwoodExpression Count, SourceSpan Span) : PipeOperation(Span);

/// <summary>| takeWhile predicate</summary>
public sealed record TakeWhileOperation(ElwoodExpression Predicate, SourceSpan Span) : PipeOperation(Span);

/// <summary>| batch n</summary>
public sealed record BatchOperation(ElwoodExpression Size, SourceSpan Span) : PipeOperation(Span);

/// <summary>| join other on leftKey equals rightKey [into alias] [inner|left|right|full]</summary>
public sealed record JoinOperation(
    ElwoodExpression Source,
    ElwoodExpression LeftKey,
    ElwoodExpression RightKey,
    string? IntoAlias,
    JoinMode Mode,
    SourceSpan Span
) : PipeOperation(Span);

public enum JoinMode { Inner, Left, Right, Full }

/// <summary>| concat separator?</summary>
public sealed record ConcatOperation(ElwoodExpression? Separator, SourceSpan Span) : PipeOperation(Span);

/// <summary>| reduce (acc, item) => expr [from initialValue]</summary>
public sealed record ReduceOperation(ElwoodExpression Accumulator, ElwoodExpression? InitialValue, SourceSpan Span) : PipeOperation(Span);

/// <summary>| any pred, | all pred</summary>
public sealed record QuantifierOperation(string Kind, ElwoodExpression Predicate, SourceSpan Span) : PipeOperation(Span);

/// <summary>| match { arms }</summary>
public sealed record MatchOperation(IReadOnlyList<MatchArm> Arms, SourceSpan Span) : PipeOperation(Span);

// ── Path Segments ──

public abstract record PathSegment(SourceSpan Span);

/// <summary>Simple property: .foo or ?.foo (optional chaining)</summary>
public sealed record PropertySegment(string Name, SourceSpan Span, bool Optional = false) : PathSegment(Span);

/// <summary>Array index: [0] or wildcard: [*]</summary>
public sealed record IndexSegment(int? Index, SourceSpan Span) : PathSegment(Span); // null = wildcard

/// <summary>Array slice: [start:end], [:end], [start:]</summary>
public sealed record SliceSegment(int? Start, int? End, SourceSpan Span) : PathSegment(Span);

/// <summary>Recursive descent: ..foo</summary>
public sealed record RecursiveDescentSegment(string Name, SourceSpan Span) : PathSegment(Span);

// ── Operators ──

public enum BinaryOperator
{
    Add, Subtract, Multiply, Divide,
    Equal, NotEqual,
    LessThan, LessThanOrEqual, GreaterThan, GreaterThanOrEqual,
    And, Or
}

public enum UnaryOperator
{
    Negate, Not
}
